storage/reads/store.go

```diff
@@ -84,7 +84,9 @@ type Store interface {
 }
 
 // WindowAggregateCapability describes what is supported by WindowAggregateStore.
-type WindowAggregateCapability struct{}
+type WindowAggregateCapability interface{}
```

Comment:
Don't you want to define the required methods for the interface here?

Reply:
Currently there is no method defined, this is something to be added later.

Reply:
It seems like to unblock Adrian's work, the methods on the interface will need to be defined, won't they?

Reply:
At this stage, as long as we see the `Reader` implements the `WindowAggregateReader` interface, it means we can push down one window aggregate function. Stabilizing the `WindowAggregateReader` interface will be enough to unblock @adrian-thurston. 

The methods in this interface will be for more detailed flags. For example, in the future we want to know if the storage engine can support pushing down *MULTIPLE* aggregates instead of one, then in addition to checking whether the `Reader` implements the `WindowAggregateReader` interface, we need to call `GetWindowAggregateCapability()` and check its `SupportMultipleAggregate()` method.

Reply:
It seems like the planner rule will want to know which aggregate functions are supported for pushdown, right?  Since only `count` will be supported initially.  It seems like we need some detail around that for the planner rule to work properly. Am I missing something?

Reply:
My understanding is that we are only going to start to use the methods in this interface after our first version goes into production (i.e., when we develop V2). I imagine the checks in the planner rule will be like

```go
func aggSupported(aggKind ProcedureKind) bool {
	if reader, ok := r.(influxdb.WindowAggregateReader); !ok {
		return false
	}
	capability := reader.GetWindowAggregateCapability()
	switch aggKind {
	case universe.CountKind:
	case **some other supported kinds in V1**:
		return true
	case universe.MeanKind:
		return capability.SupportMean()
	case **others**:
		...
	}
}
```

Reply:
Ok, I don't want to block work from proceeding.  I seem to remember some of this discussion, but I'm unclear on the details.  It seems like it would be nice if the capabilities were in place and the rule "just works" and does the right thing as new capabilities come online.

Reply:
@wolffcm @yzhang1991 The underlying communication with the server needs to have a fixed design and must be extensible. For example, a dictionary mapping names to boolean values. You can add items to the dictionary to represent additional capabilities. Newer versions of the queryd service can talk to older versions of storage. The protocol is fixed, just the data is changing.

The Go interface that the planner uses can grow and evolve as we change our needs. It just needs to use the fixed interface for communicating with storage. It can hide those details from the planner.

So for now I can interpret the fact that the reader casts to a WindowAggregateReader as indication that Count will work. Then we can update it as we go.

Reply:
On second thought ... I'll add the `HaveMin`/`HaveMax` etc functions as it will allow me to fill in the rule and test it. Then it should be ready, those functions just need implementations that flip the bit when ready.

Reply:
I'm already adding the feature flags so why not the capabilities.

Reply:
As a general rule, I try to avoid interfaces for APIs that can be or plan to be changed. The reason for this is that interfaces have some notable drawbacks in Go.

1. There is no default implementation.
2. Adding to an interface breaks code that uses it in contrast to everything else in Go.
3. When you use an interfaces sentinel null value to denote that something doesn't exist, you invite panics when a programmer doesn't realize the function can return null.

For things like capabilities, I think structs work a lot better. The reason is because you can add fields to the struct and they will automatically default to the default value for downstream code. So if we need to add a `HaveFirst` capability in the future, the downstream code that uses this struct, both producing and consuming, automatically gets the default value that it does not support the first aggregate which is correct.

The reason why I had the function return a boolean instead of the struct was because I wanted to make it possible to check the capability without assigning the value. For example, you could do this:

```
GetWindowAggregate(context.Context) (WindowAggregateCapability, bool)
```

But then you have to assign the value when you may not care what the value is as the capability doesn't have any attributes. This is why I preferred to have the capability be set as an optional argument because we didn't need it at that exact moment, but would likely need to expand it in the future.

We can potentially just switch to the above. The `master` branch didn't have the structs in a neutral location so, as long as we have a neutral location such as the `query` package, it might be fine to just keep the implementation there.